### PR TITLE
Feature: passing in optional parameters of AsyncIOMotorClient

### PR DESCRIPTION
**Pull request by dhensen**
Hi all!

(Great project! I'm already using this in my project for about 6 weeks now!)

I needed a way to pass optional parameters to AsyncIOMotorClient. I believe the solution is not the most pretties now, because there is this one configuration dict per client that is going to be cut to pieces (there was already: `connection_params`, `authMechanism`, `io_loop` and `db_name`). Any ideas on this?

The way it is now it still contains a ~bug~feature: you can override keys like username, password, host, port, authSource. I think this should not be possible though, but did build that kind of validation right now.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -38,8 +38,12 @@ check-dist:
 	python setup.py sdist
 	twine check dist/*
 
+.PHONY: start-mongo
+start-mongo:
```
Comment (i8enn) [inline Makefile]: Maybe rename this as `up-services`? I think this suits better.

File: Makefile [hunk 2]
```diff
@@ -76,3 +80,4 @@ clean:
 	rm -rf dist
 	python setup.py clean
 	rm -rf site
+	docker-compose down -v
```
Comment (i8enn) [inline Makefile]: It seems to me that it is better to put this in a separate command (as with `up-services`), adding it here as a dependency (as in tests).

Like:
```make
.PHONY: down-services
down-services:
    docker-compose down -v
...
clean: down-services
...
```

File: Makefile [hunk 3]
```diff
@@ -38,8 +38,12 @@ check-dist:
...
+.PHONY: start-mongo
+start-mongo:
+	docker-compose up -d
+
 .PHONY: test
-test:
+test: start-mongo
 	pytest --cov=pydantic_odm --cov-config=setup.cfg --no-cov-on-fail
 
 .PHONY: testwatch
```
Comment (i8enn) [inline Makefile:53]: You can add a dependency for this command as you would for the `tests` command, and remove the cyclic dependency on yourself, if not difficult?

Like:
```make
.PHONY: testwatch
testwatch: up-services
```

**Review by i8enn — CHANGES_REQUESTED** [review]
Thank you so much for contributing to this project! I am very pleased to hear that you are using my project and you like it! Thanks again!

I added some comments to the PR in review. All in all good and I think it is possible to add this to the project.

_Unfortunately, I don't understead English very well, so we may have some misunderstandings._

About:
>I believe the solution is not the most pretties now, because there is this one configuration dict per client that is going to be cut to pieces (there was already: connection_params, authMechanism, io_loop and db_name). Any ideas on this?
>
>The way it is now it still contains a bugfeature: you can override keys like username, password, host, port, authSource. I think this should not be possible though, but did build that kind of validation right now.

If I understand you correctly, you want to add immutability to the manager configuration and be able to add optional parameters for the manager itself (here I am not sure what I understood correctly)?

The first thing that came to my mind was to implement the configuration as a "pydantic" model. You can also make it possible to output additional parameters to the manager's configuration and override this in the configuration for each connection.

Something like this (very rough version):
```python
class MongoDBManagerConfiguration(BaseModel):
    sockerTimeoutMS: int = None
    connectTimeoutMS: int = None
    serverSelectionTimeoutMS: int = None
    auth_source: str: None
    databases: Dict[str, MongoDBConnectionsConfiguration] = {}

class MongoDBConnectionsConfiguration(BaseModel):
    username: str
    password: str
    auth_source: str
    ...
```

I think this should be taken out as a separate issue and discussed there.

How do you look at it?

_Thanks again for your contribution!_

File: Makefile (hunk 3 above)
Comment (dhensen) [reply Makefile:53]: Thanks for reviewing, I will fix it this week!

Comment (dhensen) [reply Makefile:53]: Fixed it 👍 I did not have encounter any more issues, I could basically copy-paste what you proposed.

**Comment by dhensen** [comment]
> Thank you so much for contributing to this project! I am very pleased to hear that you are using my project and you like it! Thanks again!
> 
> I added some comments to the PR in review. All in all good and I think it is possible to add this to the project.
> 
> _Unfortunately, I don't understead English very well, so we may have some misunderstandings._
> 
> About:
> 
> > I believe the solution is not the most pretties now, because there is this one configuration dict per client that is going to be cut to pieces (there was already: connection_params, authMechanism, io_loop and db_name). Any ideas on this?
> > The way it is now it still contains a bugfeature: you can override keys like username, password, host, port, authSource. I think this should not be possible though, but did build that kind of validation right now.
> 
> If I understand you correctly, you want to add immutability to the manager configuration and be able to add optional parameters for the manager itself (here I am not sure what I understood correctly)?
> 
> The first thing that came to my mind was to implement the configuration as a "pydantic" model. You can also make it possible to output additional parameters to the manager's configuration and override this in the configuration for each connection.
> 
> Something like this (very rough version):
> 
> ```python
> class MongoDBManagerConfiguration(BaseModel):
>     sockerTimeoutMS: int = None
>     connectTimeoutMS: int = None
>     serverSelectionTimeoutMS: int = None
>     auth_source: str: None
>     databases: Dict[str, MongoDBConnectionsConfiguration] = {}
> 
> class MongoDBConnectionsConfiguration(BaseModel):
>     username: str
>     password: str
>     auth_source: str
>     ...
> ```
> 
> I think this should be taken out as a separate issue and discussed there.
> 
> How do you look at it?
> 
> _Thanks again for your contribution!_

I did not mean to want to have it be immutable. What I meant was, when you look at the [mongoClient docs](https://pymongo.readthedocs.io/en/3.11.0/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient), you see a list op optional parameters. That list contains things like socketTimeoutMS, connectTimeoutMS, etc... but also username, password, authSource, host, post. These last five fields are already supported by `pydantic-odm`. Basically the AsyncIOMotorClient/MongoClient already has a configuration API: the kwargs in the constructor. Because pydantic-odm is now wrapping AsyncIOMotorClient, we need to think how we want to expose other fields that are maybe not so commonly used. One way is as we did in this PR: through the use of `OPTIONAL_PARAMETERS`.

But now we have cut the configuration in two pieces: we need to make one dictionary containing username, password, host, port, authSource. And an optional second dictionary containing other fields. It works now, but I feel more for adapting to API of the original AsyncIOMotorClient constructor: passing them all in via the same kwargs.

Then indeed you can use your proposed:
```python
class MongoDBConnectionsConfiguration(BaseModel):
    username: str
    password: str
    auth_source: str
    socketTimeoutMS: int
    connectTimeoutMS: int
    ...
```

But you're right, let's take this to another issue!

**Review by i8enn — DISMISSED** [review]
Excellent! Thanks for your fixes!

**Comment by i8enn** [comment]
@dhensen I'm sorry, I really didn't fully understand you!

> I did not mean to want to have it be immutable. What I meant was, when you look at the mongoClient docs, you see a list op optional parameters. That list contains things like socketTimeoutMS, connectTimeoutMS, etc... but also username, password, authSource, host, post. These last five fields are already supported by pydantic-odm. Basically the AsyncIOMotorClient/MongoClient already has a configuration API: the kwargs in the constructor. Because pydantic-odm is now wrapping AsyncIOMotorClient, we need to think how we want to expose other fields that are maybe not so commonly used. One way is as we did in this PR: through the use of OPTIONAL_PARAMETERS.
> 
> But now we have cut the configuration in two pieces: we need to make one dictionary containing username, password, host, port, authSource. And an optional second dictionary containing other fields. It works now, but I feel more for adapting to API of the original AsyncIOMotorClient constructor: passing them all in via the same kwargs.

Now I understand what you meant.

This is actually very simple: we just allow optional parameters to be entered into the same dictionary as the main ones - then we send them to the AsyncIOMotorClient / MongoClient `kwargs` and everything works.

But it would not works. Unfortunately, I decided to name the parameters a little "pytonic", because of this they have to be converted to the AsyncIOMotorClient / MongoClient format. It's ok if there are not many parameters, but in our case optional parameters are rare and there are a lot of them.

As a result, I have a problem:
- either I name the parameters as in AsyncIOMotorClient / MongoClient and disclaim responsibility for maintaining them (I just pass them into the constructor);
- or I make the names of the AsyncIOMotorClient / MongoClient parameters more "pythonic"

Of course, the first option is more practical, but I'm not very happy to see camelCase in the config file - although it may be PEP-paranoia.

And what do you think?

**Comment by i8enn** [comment]
@dhensen I have some ideas on how to do this. Do you mind if I add this to your PR?

**Comment by dhensen** [comment]
Hi Ivan,

Go ahead! (My preference would be to have it pythonic, but I also feel like
having it be 1-to-1 with the wrapped libraries... this is very
inconclusive, haha what a dilemma)

**Review by i8enn — DISMISSED** [review]

**Comment by i8enn** [comment]
> Hi Ivan, Go ahead! (My preference would be to have it pythonic, but I also feel like having it be 1-to-1 with the wrapped libraries... this is very inconclusive, haha what a dilemma)

@dhensen okay, let's leave this option for now.

**Review by i8enn — APPROVED** [review]

**Pull request merged**